File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/projecting/MatchedProjector.java
package com.codepoetics.raffia.streaming.projecting;

import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

final class MatchedProjector<T extends BasketWriter<T>> extends StreamingProjector<T> {

  private int depth = 0;

  MatchedProjector(T target, FilteringWriter<T> parent) {
    super(target, parent);
    this.parent = parent;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    target = target.beginObject();
    depth += 1;
    return this;
  }

  @Override
  public FilteringWriter<T> beginArray() {
    target = target.beginArray();
    depth +=1;
    return this;
  }

  @Override
  public FilteringWriter<T> end() {
    target = target.end();

    if (depth-- > 0) {
      return this;
    }

    if (parent == null) {
      throw new IllegalStateException("end() called when not writing object or array");
    }

    return parent.advance(target);
  }

  @Override
  public FilteringWriter<T> key(String key) {
    target = target.key(key);
    return this;
  }

  @Override
  public FilteringWriter<T> add(String value) {
    target = target.add(value);
    return this;
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    target = target.add(value);
    return this;
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    target = target.add(value);
    return this;
  }

  @Override
  public FilteringWriter<T> addNull() {
    target = target.addNull();
    return this;
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/projecting/MatchedProjector.kt
package com.codepoetics.raffia.streaming.projecting

import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

internal class MatchedProjector<T : BasketWriter<T>>(target: T, parent: FilteringWriter<T>?) : StreamingProjector<T>(target, parent) {

    private var depth = 0

    override fun beginObject(): FilteringWriter<T> {
        target = target.beginObject()
        depth += 1
        return this
    }

    override fun beginArray(): FilteringWriter<T> {
        target = target.beginArray()
        depth += 1
        return this
    }

    override fun end(): FilteringWriter<T> {
        target = target.end()

        if (depth-- > 0) {
            return this
        }

        if (parent == null) {
            throw IllegalStateException("end() called when not writing object or array")
        }

        return parent.advance(target)
    }

    override fun key(key: String): FilteringWriter<T> {
        target = target.key(key)
        return this
    }

    override fun add(value: String): FilteringWriter<T> {
        target = target.add(value)
        return this
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        target = target.add(value)
        return this
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        target = target.add(value)
        return this
    }

    override fun addNull(): FilteringWriter<T> {
        target = target.addNull()
        return this
    }
}
